mantidproject/mantid
# Extend IFunction for attributes with a Validator

**Description of work.**

The function interface IFunction has been updated to allow for the declaration of function attributes with a validator. Such a validator provides a method to restrict the possible value of the attribute. The following validators (previously implemented for algorithm parameters) have been tested:
- String List Validator
- BoundedValidator
- ArrayBoundedValidator
- StringContainsValidator

It is of note that the value of attributes can be changed using attribute visitors, for which a pure virtual interface is provided in IFunction. Through this pure virtual interface the implementation of attribute visitors is specific to individual functions. Is is therefore the responsibility of the author of those functions to ensure attribute visitors are implemented so validators are evaluated upon change of attribute value (using functions provided in IFunction).

**To test:**

This change is to add additional functionality to the C++ IFunction Interace. As such, this additional functionality cannot be directly tested via a workbench interface or through running a script using the python API (though through such testing you could check for regression).

- In order to test the new functionality, one could observe the unit test, "FunctionAttributeValidatorTest", provided.

- Alternatively, choose an existing *C++* function that declares an attribute and add a validator (from the above list) to this 
  attribute so that:
  `void IFunction::declareAttribute(const std::string &name, const API::IFunction::Attribute &defaultValue,
                                 const Kernel::IValidator &validator)`
  As opposed to solely:
  `void IFunction::declareAttribute(const std::string &name, const API::IFunction::Attribute &defaultValue)`

  The syntax for creating validators can be found in the unit test, "FunctionAttributeValidatorTest".

- Check that given a validator, the attribute cannot be set to invalid values.



Resolves #33120

*This does not require release notes* because this is one of a series of related PRs, release note will be added in concluding PR.

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
